src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -128,12 +128,76 @@ public override void VisitIfConditionSyntax(IfConditionSyntax syntax)
                 return this.typeManager.GetTypeInfo(syntax.Body);
             });
 
+        public override void VisitLocalVariableSyntax(LocalVariableSyntax syntax)
+            => AssignType(syntax, () =>
+            {
+                var parent = this.binder.GetParent(syntax);
+                switch (parent)
+                {
+                    case ForSyntax @for when ReferenceEquals(syntax, @for.ItemVariable):
+                        // this local variable is a loop item variable
+                        // we should return item type of the array (if feasible)
+
+                        // get type of the loop array expression
+                        // (this shouldn't cause a stack overflow because it's a peer node of this one)
+                        var arrayExpressionType = this.typeManager.GetTypeInfo(@for.Expression);
+
+                        if (arrayExpressionType.TypeKind == TypeKind.Any || arrayExpressionType is not ArrayType arrayType)
+                        {
+                            // the array is of "any" type or the loop array expression isn't actually an array
+                            // in the former case, there isn't much we can do
+                            // in the latter case, we will let the ForSyntax type check rules produce the error for it
+                            return LanguageConstants.Any;
+                        }
+
+                        // the array expression is actually an array
+                        return arrayType.Item;
+
+                    default:
+                        throw new InvalidOperationException($"{syntax.GetType().Name} at {syntax.Span} has an unexpected parent of type {parent?.GetType().Name}");
+                }
+            });
+
+        public override void VisitForSyntax(ForSyntax syntax)
+            => AssignType(syntax, () =>
+            {
+                var errors = new List<ErrorDiagnostic>();
+
+                var loopItemType = typeManager.GetTypeInfo(syntax.ItemVariable);
+                CollectErrors(errors, loopItemType);
+
+                var arrayExpressionType = typeManager.GetTypeInfo(syntax.Expression);
+                CollectErrors(errors, arrayExpressionType);
+
+                var bodyType = typeManager.GetTypeInfo(syntax.Body);
+                CollectErrors(errors, bodyType);
+
+                if (PropagateErrorType(errors, loopItemType, arrayExpressionType, bodyType))
+                {
+                    return ErrorType.Create(errors);
+                }
+
+                if (!TypeValidator.AreTypesAssignable(arrayExpressionType, LanguageConstants.Array))
+                {
+                    // the array expression isn't actually an array

```

Comment:
Does this account for invalid type definitions, and show a warning instead of an error as appropriate?

Reply:
So assigning a for-expression to a property whose type isn't array will generate a warning. However, referencing a non-array in a loop array expression currently produces an error. Potentially, the user could reference a mis-typed property via property access, but I'd rather leave this strict for now and see if we get complaints.

There's also a separate issue if I can even allow non deploy-time constants here anyway. I'll have to investigate that one.